src/test/basics/PerfLog_test.cpp

```diff
@@ -0,0 +1,1075 @@
+            parent.doStart();
+            std::this_thread::sleep_for (parent.getLogInterval() * 3);
+            parent.doStop();
+
+            // Remove the file.
+            remove (perfLogPath);
+        }
+        {
+            // Put a write protected file where PerfLog wants to write its
+            // file.  Make sure that PerfLog tries to shutdown the server
+            // since it can't open its file.
+            PerfLogParent parent {j_};
+            if (! BEAST_EXPECT(! exists (perfLogPath)))
+                return;
+
+            // Construct and write protect a file to prevent PerfLog
+            // from creating its file.
+            boost::system::error_code ec;
+            boost::filesystem::create_directories (perfLogPath, ec);
+            if (! BEAST_EXPECT(! ec))
+                return;
+
+            std::ofstream woFile;
+            woFile.open (fullPath.c_str(), std::ios::out | std::ios::app);
+            if (! BEAST_EXPECT(woFile))
+                return;
+            woFile.close();
+            permissions (fullPath,
+                perms::remove_perms | perms::owner_write |
+                perms::others_write | perms::group_write);
+
+            // Now construct a PerfLog.  The PerfLog should attempt to shut
+            // down the server because it can't open its file.
+            BEAST_EXPECT(parent.stopSignaled == false);
+            auto perfLog {getPerfLog (parent, WithFile::yes)};
+            BEAST_EXPECT(parent.stopSignaled == true);
+
+            // Start PerfLog and wait long enough for PerfLog::report()
+            // to not be able to write to its file.  That should cause no
+            // problems.
+            parent.doStart();
+            std::this_thread::sleep_for (parent.getLogInterval() * 3);
+            parent.doStop();
+
+            // Fix file permissions so the file can be cleaned up.
+            permissions (fullPath,
+                perms::add_perms | perms::owner_write |
+                perms::others_write | perms::group_write);
+        }
+    }
+
+    void testRPC (WithFile withFile)
+    {
+        // Exercise the rpc interfaces of PerfLog.
+        // Start up the PerfLog that we'll use for testing.
+        PerfLogParent parent {j_};
+        auto perfLog {getPerfLog (parent, withFile)};
+        parent.doStart();
+
+        // Get the all the labels we can use for RPC interfaces without
+        // causing an assert.
+        std::vector<char const*> labels {ripple::RPC::getHandlerNames()};
+        std::shuffle (labels.begin(), labels.end(), shuffler_);
+
+        // Get two IDs to associate with each label.  Errors tend to happen at
+        // boundaries, so we pick IDs starting from zero and ending at
+        // std::uint64_t>::max().
+        std::vector<std::uint64_t> ids;
+        ids.reserve (labels.size() * 2);
+        std::generate_n (std::back_inserter (ids), labels.size(),
+            [i = std::numeric_limits<std::uint64_t>::min()]()
```

Comment:
Why `std::numeric_limits<...>::min()` for an unsigned type? Maybe `i = std::uint64_t(0)` or somesuch?

Reply:
That's fair.  It's just preserving symmetry with the `numeric_limits::max()` on line 336.